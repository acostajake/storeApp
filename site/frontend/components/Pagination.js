import Head from 'next/head';
import Link from 'next/link';
import React from 'react';
import { Query } from 'react-apollo';

import { perPage } from '../config';
import PaginationStyles from './styles/PaginationStyles';
import { PAGINATION_QUERY } from '../lib/queries';

const Pagination = (props) => {
	return (
		<Query query={PAGINATION_QUERY}>
			{({ data, loading, error }) => {
				if (loading) return <p>Loading pages...</p>;
				const count = data.itemsConnection.aggregate.count;
				const pages = Math.ceil(count / perPage);
				const page = props.page;
				return (
					<PaginationStyles>
						<Head>
							<title>Store | Page {page}</title>
						</Head>
						<Link
							prefetch
							href={{ pathname: 'items', query: { page: page - 1 } }}
						>
							<a className='prev' aria-disabled={page <= 1}>
								Prev
							</a>
						</Link>
						<p>
							Page {page} of {pages}
						</p>
						<p>{count} items total</p>
						<Link
							prefetch
							href={{ pathname: 'items', query: { page: page + 1 } }}
						>
							<a className='prev' aria-disabled={page >= pages}>
								Next
							</a>
						</Link>
					</PaginationStyles>
				);
			}}
		</Query>
	);
};

export default Pagination;
