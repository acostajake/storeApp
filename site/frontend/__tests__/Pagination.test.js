import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';

import {
	makeMocksFor,
	notSignedInMocks,
	signedInMocks,
} from '../lib/testUtils';
import Pagination from '../components/Pagination';

Router.router = {
	push() {},
	prefetch() {},
};

describe('<Pagination/>', () => {
	it('displays a loading message', () => {
		const mountPagination = mount(
			<MockedProvider mocks={makeMocksFor(3)}>
				<Pagination />
			</MockedProvider>
		);
		expect(mountPagination.text()).toContain('Loading pages...');
	});
	it('renders pagination for 13 items', async () => {
		const mountPagination = mount(
			<MockedProvider mocks={makeMocksFor(13)}>
				<Pagination page={1} />
			</MockedProvider>
		);
		await wait();
		mountPagination.update();
		const pagination = mountPagination.find('[data-test="pagination"]');
		expect(toJSON(pagination)).toMatchSnapshot();
		expect(mountPagination.find('.totalPages').text()).toBe('4');
	});
	it('disables prev button on first page', async () => {
		const mountPagination = mount(
			<MockedProvider mocks={makeMocksFor(13)}>
				<Pagination page={1} />
			</MockedProvider>
		);
		await wait();
		mountPagination.update();
		const prev = mountPagination.find('a.prev');
		expect(prev.prop('aria-disabled')).toEqual(true);
	});
	it('disables next button on last page', async () => {
		const mountPagination = mount(
			<MockedProvider mocks={makeMocksFor(13)}>
				<Pagination page={4} />
			</MockedProvider>
		);
		await wait();
		mountPagination.update();
		const next = mountPagination.find('a.next');
		expect(next.prop('aria-disabled')).toEqual(true);
	});
	it('enables all buttons on middle page', async () => {
		const mountPagination = mount(
			<MockedProvider mocks={makeMocksFor(13)}>
				<Pagination page={2} />
			</MockedProvider>
		);
		await wait();
		mountPagination.update();
		const next = mountPagination.find('a.next');
		const prev = mountPagination.find('a.prev');
		expect(next.prop('aria-disabled')).toEqual(false);
		expect(prev.prop('aria-disabled')).toEqual(false);
	});
});
