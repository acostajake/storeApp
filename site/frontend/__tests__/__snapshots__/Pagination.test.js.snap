// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> renders pagination for 13 items 1`] = `
Array [
  <PaginationStyles
    data-test="pagination"
  >
    <div
      className="PaginationStyles-sc-7n6p6m-0 hOgZr"
      data-test="pagination"
    >
      <SideEffect(Head)>
        <Head />
      </SideEffect(Head)>
      <Link
        href={
          Object {
            "pathname": "items",
            "query": Object {
              "page": 0,
            },
          }
        }
        prefetch={true}
      >
        <a
          aria-disabled={true}
          className="prev"
          href="items?page=0"
          onClick={[Function]}
        >
          Prev
        </a>
      </Link>
      <p>
        Page 
        1
         of 
        <span
          className="totalPages"
        >
          4
        </span>
      </p>
      <p>
        13
         items total
      </p>
      <Link
        href={
          Object {
            "pathname": "items",
            "query": Object {
              "page": 2,
            },
          }
        }
        prefetch={true}
      >
        <a
          aria-disabled={false}
          className="next"
          href="items?page=2"
          onClick={[Function]}
        >
          Next
        </a>
      </Link>
    </div>
  </PaginationStyles>,
  <div
    className="PaginationStyles-sc-7n6p6m-0 hOgZr"
    data-test="pagination"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <Link
      href={
        Object {
          "pathname": "items",
          "query": Object {
            "page": 0,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={true}
        className="prev"
        href="items?page=0"
        onClick={[Function]}
      >
        Prev
      </a>
    </Link>
    <p>
      Page 
      1
       of 
      <span
        className="totalPages"
      >
        4
      </span>
    </p>
    <p>
      13
       items total
    </p>
    <Link
      href={
        Object {
          "pathname": "items",
          "query": Object {
            "page": 2,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={false}
        className="next"
        href="items?page=2"
        onClick={[Function]}
      >
        Next
      </a>
    </Link>
  </div>,
]
`;
